File: repair-room-once.mjs
// Bounded operator repair, not AgentTeams runtime evidence.
// Only: read team/membership -> pause p5 with readback -> send one corrected p6.
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { pathToFileURL } from 'node:url';

export const OLD_ID = 'cloud-crisis-2014-east-art-center-p5';
export const NEW_ID = 'cloud-crisis-2014-east-art-center-p6';
export const TEAM = 'juchang-change-control-v14';
const roles = ['material_intake', 'evidence_guard', 'entity_matcher', 'approval_guard'];
const deps = [[], [0], [0], [1, 2]];
const digest = value => crypto.createHash('sha256').update(JSON.stringify(value)).digest('hex');
const ensure = (ok, code) => { if (!ok) throw new Error(code); };

export function correctedEnvelope(snapshot, team) {
  const old = snapshot.request;
  ensure(old?.schema === 'juchang-agentteams-dsh-project@1' && old.projectId === OLD_ID, 'OLD_REQUEST_ID_MISMATCH');
  ensure(old.publicWriteAllowed === false && old.inputPayload && typeof old.inputPayload === 'object', 'OLD_INPUT_OR_AUTHORITY_INVALID');
  ensure(!old.studio && !old.failureDrill, 'UNSUPPORTED_OLD_RUN_CONTEXT');
  ensure(team.name === TEAM && team.phase === 'Active' && team.leaderReady === true && team.readyWorkers === 4, 'TEAM_NOT_READY');
  ensure(/^![^:\s]+:.+$/.test(team.teamRoomID || '') && /^![^:\s]+:.+$/.test(team.leaderDMRoomID || ''), 'ROOM_FIELDS_MISSING');
  ensure(team.teamRoomID !== team.leaderDMRoomID && old.roomId === team.leaderDMRoomID, 'ROOM_REPAIR_SCOPE_MISMATCH');
  ensure(snapshot.failure?.message?.includes('is not a joined member of room') && snapshot.failure.message.includes(old.roomId), 'OLD_FAILURE_MISMATCH');
  const audit = snapshot.audit?.events || [];
  ensure(audit.some(e => e.operation === 'create_project' && e.success === true) && audit.some(e => e.operation === 'plan_dag' && e.success === true), 'OLD_PLAN_NOT_OBSERVED');
  ensure(!audit.some(e => ['delegate_task', 'accept_task_result', 'complete_project'].includes(e.operation) && e.success === true), 'OLD_RUN_ADVANCED_STOP');
  ensure(Array.isArray(old.tasks) && old.tasks.length === 4, 'OLD_TASK_COUNT_INVALID');
  const tasks = old.tasks.map((task, i) => {
    ensure(task.role === roles[i] && task.taskId === `${OLD_ID}-0${i + 1}`, 'OLD_TASK_IDENTITY_INVALID');
    ensure(/^@[^:\s]+:.+$/.test(task.assignedTo || ''), 'FULL_ASSIGNEE_ID_REQUIRED');
    ensure(JSON.stringify(task.dependsOn) === JSON.stringify(deps[i].map(j => `${OLD_ID}-0${j + 1}`)), 'OLD_DAG_INVALID');
    return { taskId: `${NEW_ID}-0${i + 1}`, role: task.role, title: task.title, assignedTo: task.assignedTo,
      dependsOn: deps[i].map(j => `${NEW_ID}-0${j + 1}`) };
  });
  return {
    schema: old.schema, dispatchId: NEW_ID, projectRef: old.projectRef, title: old.title,
    intakeKind: old.intakeKind, sourceUrl: old.sourceUrl, sourceAuthor: old.sourceAuthor,
    sourceHash: old.sourceHash, roomId: team.teamRoomID, publicWriteAllowed: false,
    inputPayload: old.inputPayload, tasks,
  };
}

export async function repairOnce(io, snapshot, apply = false) {
  const team = await io.team();
  const envelope = correctedEnvelope(snapshot, team);
  const leader = await io.leader(team.leaderName);
  ensure(/^@[^:\s]+:.+$/.test(leader || ''), 'LEADER_ID_INVALID');
  const members = new Set(await io.members(team.teamRoomID));
  ensure([leader, ...envelope.tasks.map(t => t.assignedTo)].every(id => members.has(id)), 'TEAM_JOINED_MEMBERSHIP_FAILED');
  const dm = new Set(await io.members(team.leaderDMRoomID));
  ensure(dm.has(leader) && dm.has(io.senderId), 'DM_MEMBERSHIP_FAILED');
  ensure(await io.project(NEW_ID, true) === null, 'NEW_PROJECT_ALREADY_EXISTS_NO_RESEND');
  const old = await io.project(OLD_ID);
  ensure(['active', 'paused'].includes(old.status), 'OLD_PROJECT_STATE_UNSUPPORTED');
  const nodes = old.nodes || old.tasks || [];
  ensure(Array.isArray(nodes) && nodes.length === 4, 'OLD_OFFICIAL_DAG_UNEXPECTED');
  const expectedIds = new Set(snapshot.request.tasks.map(t => t.taskId));
  ensure(nodes.every(n => expectedIds.has(n.id || n.task_id || n.taskId)), 'OLD_OFFICIAL_TASKS_MISMATCH');
  ensure(!nodes.some(n => ['in_progress', 'submitted', 'completed'].includes(n.status)), 'OLD_OFFICIAL_RUN_ADVANCED_STOP');
  const plan = { schema: 'juchang-room-repair-plan@1', oldId: OLD_ID, newId: NEW_ID,
    roomsSeparated: true, joinedWorkers: 4, payloadSha256: digest(envelope.inputPayload),
    envelopeSha256: digest(envelope), productionWrites: 0, publicPublishes: 0, realRefunds: 0, externalBusinessMessages: 0 };
  io.print('PRECHECK_OK rooms_separated=true joined_workers=4');
  if (!apply) return plan;
  // A durable attempt marker is claimed BEFORE any mutation. Uncertain attempts never auto-repeat.
  io.claim(plan);
  if (old.status !== 'paused') await io.pause(OLD_ID);
  ensure((await io.project(OLD_ID)).status === 'paused', 'P5_PAUSE_READBACK_FAILED');
  io.print('P5_PAUSED_READBACK_OK');
  io.record('prepared-envelope.json', envelope);
  const body = {
    msgtype: 'm.text',
    body: `JUCHANG room-corrected run. Preserve p5 failure; no business writes.\n${leader}\nJUCHANG_DSH_PROJECT: ${JSON.stringify(envelope)}`,
    'm.mentions': { user_ids: [leader] },
  };
  const sent = await io.send(team.leaderDMRoomID, `juchang-roomfix-${NEW_ID}-once`, body);
  ensure(typeof sent.event_id === 'string' && sent.event_id.startsWith('$'), 'SEND_RESULT_UNKNOWN_DO_NOT_REPEAT');
  io.record('sent.json', { ...plan, sentAt: new Date().toISOString(), eventId: sent.event_id,
    state: 'sent_not_completed', instruction: 'Read exact p6 Project/Taskflow; never infer completion from this record.' });
  io.print(`DISPATCH_SENT project=${NEW_ID} event_prefix=${sent.event_id.slice(0, 12)}`);
  io.print('NOT_COMPLETED: wait for official four-task and Leader terminal readback.');
  return plan;
}

function cli(args, token, allow404 = false) {
  try {
    const out = execFileSync('agt', args, { encoding: 'utf8', timeout: 35000,
      env: { ...process.env, ...(token ? { AGENTTEAMS_AUTH_TOKEN: token } : {}) },
      stdio: ['ignore', 'pipe', 'pipe'], maxBuffer: 2 * 1024 * 1024 });
    return JSON.parse(out);
  } catch (e) {
    if (allow404 && /HTTP 404\b/.test(String(e.stderr || ''))) return null;
    throw new Error(`AGT_${args[0].toUpperCase()}_${args[1].toUpperCase()}_FAILED`);
  }
}

async function main() {
  const [mode, snapshotFile, stateDir] = process.argv.slice(2);
  ensure(['--plan', '--apply'].includes(mode) && snapshotFile && stateDir, 'USAGE_PLAN_OR_APPLY_SNAPSHOT_STATE_DIR');
  const snapshot = JSON.parse(fs.readFileSync(snapshotFile, 'utf8'));
  fs.mkdirSync(stateDir, { recursive: true, mode: 0o700 });
  if (mode === '--apply' && fs.existsSync(path.join(stateDir, 'attempt.json'))) throw new Error('ATTEMPT_EXISTS_NO_RESEND');
  const base = 'http://127.0.0.1:6167';
  let token = '';
  async function api(method, endpoint, body, auth = true) {
    let response;
    try {
      response = await fetch(base + endpoint, { method,
        headers: { 'Content-Type': 'application/json', ...(auth ? { Authorization: `Bearer ${token}` } : {}) },
        ...(body === undefined ? {} : { body: JSON.stringify(body) }), signal: AbortSignal.timeout(15000) });
    } catch { throw new Error('MATRIX_REQUEST_UNCERTAIN_DO_NOT_REPEAT'); }
    ensure(response.ok, `MATRIX_HTTP_${response.status}`);
    try { return await response.json(); } catch { throw new Error('MATRIX_RESPONSE_INVALID'); }
  }
  try {
    ensure(Boolean(process.env.AGENTTEAMS_ADMIN_PASSWORD), 'SERVER_ADMIN_CREDENTIAL_UNAVAILABLE');
    const login = await api('POST', '/_matrix/client/v3/login', { type: 'm.login.password',
      identifier: { type: 'm.id.user', user: 'admin' }, password: process.env.AGENTTEAMS_ADMIN_PASSWORD }, false);
    token = login.access_token;
    ensure(typeof token === 'string' && token.length > 0, 'MATRIX_LOGIN_FAILED');
    const who = await api('GET', '/_matrix/client/v3/account/whoami');
    ensure(/^@admin:/.test(who.user_id || ''), 'SENDER_NOT_ADMIN');
    const io = {
      senderId: who.user_id,
      team: async () => cli(['get', 'teams', TEAM, '-o', 'json'], token),
      leader: async name => cli(['get', 'workers', name, '-o', 'json'], token).matrixUserID,
      members: async room => (await api('GET', `/_matrix/client/v3/rooms/${encodeURIComponent(room)}/members?membership=join`)).chunk
        .filter(e => e.type === 'm.room.member' && e.content?.membership === 'join').map(e => e.state_key),
      project: async (id, optional = false) => cli(['get', 'projects', id, '-o', 'json'], token, optional),
      pause: async id => cli(['project', 'pause', id, '--reason', 'Authorized room mismatch repair; preserve p5 evidence'], token),
      claim: value => fs.writeFileSync(path.join(stateDir, 'attempt.json'), JSON.stringify({ ...value, at: new Date().toISOString() }), { flag: 'wx', mode: 0o600 }),
      record: (name, value) => fs.writeFileSync(path.join(stateDir, name), JSON.stringify(value, null, 2), { flag: 'wx', mode: 0o600 }),
      send: (room, txn, body) => api('PUT', `/_matrix/client/v3/rooms/${encodeURIComponent(room)}/send/m.room.message/${encodeURIComponent(txn)}`, body),
      print: line => console.log(line),
    };
    await repairOnce(io, snapshot, mode === '--apply');
    if (mode === '--apply') {
      for (let i = 0; i < 6; i++) {
        await new Promise(resolve => setTimeout(resolve, 3000));
        let project;
        try { project = await io.project(NEW_ID, true); }
        catch { console.log('READBACK_UNAVAILABLE_SENT_NOT_RETRIED'); break; }
        if (!project) continue;
        const nodes = project.nodes || project.tasks || [];
        const states = nodes.map(n => ({ task: String(n.id || n.task_id || n.taskId || '').slice(-2), status: n.status }));
        console.log('P6_OFFICIAL_READBACK ' + JSON.stringify({ status: project.status, tasks: states }));
        if (nodes.some(n => ['assigned', 'in_progress', 'submitted', 'completed'].includes(n.status))) break;
      }
    }
  } finally {
    if (token) { try { await api('POST', '/_matrix/client/v3/logout', {}); } catch { console.log('SESSION_LOGOUT_UNCONFIRMED'); } }
    token = '';
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href) {
  main().catch(e => {
    const message = String(e.message || '');
    console.error('STOP: ' + (/^[A-Z0-9_]+$/.test(message) ? message : 'LOCAL_OPERATION_FAILED'));
    process.exitCode = 1;
  });
}

File: repair-room-once.py
#!/usr/bin/env python3
"""Bounded operator repair. Not AgentTeams runtime evidence. Stdlib only."""
import copy
import datetime
import hashlib
import json
import os
from pathlib import Path
import re
import subprocess
import sys
import time
import urllib.error
import urllib.parse
import urllib.request

OLD_ID = "cloud-crisis-2014-east-art-center-p5"
NEW_ID = "cloud-crisis-2014-east-art-center-p6"
TEAM = "juchang-change-control-v14"
ROLES = ["material_intake", "evidence_guard", "entity_matcher", "approval_guard"]
DEPS = [[], [0], [0], [1, 2]]


class RepairError(Exception):
    """Only safe constant codes may reach stdout/stderr."""


def ensure(ok, code):
    if not ok:
        raise RepairError(code)


def encode(value):
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"), allow_nan=False)


def digest(value):
    return hashlib.sha256(encode(value).encode("utf-8")).hexdigest()


def now():
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


def corrected_envelope(snapshot, team):
    old = snapshot.get("request", {})
    ensure(old.get("schema") == "juchang-agentteams-dsh-project@1" and old.get("projectId") == OLD_ID, "OLD_REQUEST_ID_MISMATCH")
    ensure(old.get("publicWriteAllowed") is False and isinstance(old.get("inputPayload"), dict), "OLD_INPUT_OR_AUTHORITY_INVALID")
    ensure(not old.get("studio") and not old.get("failureDrill"), "UNSUPPORTED_OLD_RUN_CONTEXT")
    ensure(team.get("name") == TEAM and team.get("phase") == "Active" and team.get("leaderReady") is True and team.get("readyWorkers") == 4, "TEAM_NOT_READY")
    ensure(all(re.fullmatch(r"![^:\s]+:.+", team.get(k, "")) for k in ("teamRoomID", "leaderDMRoomID")), "ROOM_FIELDS_MISSING")
    ensure(team["teamRoomID"] != team["leaderDMRoomID"] and old.get("roomId") == team["leaderDMRoomID"], "ROOM_REPAIR_SCOPE_MISMATCH")
    message = snapshot.get("failure", {}).get("message", "")
    ensure("is not a joined member of room" in message and old["roomId"] in message, "OLD_FAILURE_MISMATCH")
    audit = snapshot.get("audit", {}).get("events", [])
    ensure(all(any(e.get("operation") == op and e.get("success") is True for e in audit) for op in ("create_project", "plan_dag")), "OLD_PLAN_NOT_OBSERVED")
    ensure(not any(e.get("operation") in ("delegate_task", "accept_task_result", "complete_project") and e.get("success") is True for e in audit), "OLD_RUN_ADVANCED_STOP")
    ensure(isinstance(old.get("tasks"), list) and len(old["tasks"]) == 4, "OLD_TASK_COUNT_INVALID")
    tasks = []
    for i, task in enumerate(old["tasks"]):
        ensure(task.get("role") == ROLES[i] and task.get("taskId") == "{}-0{}".format(OLD_ID, i + 1), "OLD_TASK_IDENTITY_INVALID")
        ensure(re.fullmatch(r"@[^:\s]+:.+", task.get("assignedTo", "")), "FULL_ASSIGNEE_ID_REQUIRED")
        ensure(task.get("dependsOn") == ["{}-0{}".format(OLD_ID, j + 1) for j in DEPS[i]], "OLD_DAG_INVALID")
        replacement = {k: task[k] for k in ("role", "title", "assignedTo") if k in task}
        replacement.update(taskId="{}-0{}".format(NEW_ID, i + 1), dependsOn=["{}-0{}".format(NEW_ID, j + 1) for j in DEPS[i]])
        tasks.append(replacement)
    envelope = {k: copy.deepcopy(old[k]) for k in ("schema", "projectRef", "title", "intakeKind", "sourceUrl", "sourceAuthor", "sourceHash", "inputPayload") if k in old}
    envelope.update(dispatchId=NEW_ID, roomId=team["teamRoomID"], publicWriteAllowed=False, tasks=tasks)
    return envelope


def repair_once(io, snapshot, apply=False):
    team = io.team()
    envelope = corrected_envelope(snapshot, team)
    leader = io.leader(team["leaderName"])
    ensure(isinstance(leader, str) and re.fullmatch(r"@[^:\s]+:.+", leader), "LEADER_ID_INVALID")
    members = set(io.members(team["teamRoomID"]))
    ensure(all(mxid in members for mxid in [leader] + [t["assignedTo"] for t in envelope["tasks"]]), "TEAM_JOINED_MEMBERSHIP_FAILED")
    dm = set(io.members(team["leaderDMRoomID"]))
    ensure(leader in dm and io.sender_id in dm, "DM_MEMBERSHIP_FAILED")
    ensure(io.project(NEW_ID, optional=True) is None, "NEW_PROJECT_ALREADY_EXISTS_NO_RESEND")
    old = io.project(OLD_ID)
    ensure(old.get("status") in ("active", "paused"), "OLD_PROJECT_STATE_UNSUPPORTED")
    nodes = old.get("nodes", old.get("tasks", []))
    ensure(isinstance(nodes, list) and len(nodes) == 4, "OLD_OFFICIAL_DAG_UNEXPECTED")
    expected = {t["taskId"] for t in snapshot["request"]["tasks"]}
    ensure({n.get("id", n.get("task_id", n.get("taskId"))) for n in nodes} == expected, "OLD_OFFICIAL_TASKS_MISMATCH")
    ensure(not any(n.get("status") in ("in_progress", "submitted", "completed") for n in nodes), "OLD_OFFICIAL_RUN_ADVANCED_STOP")
    plan = dict(schema="juchang-room-repair-plan@1", oldId=OLD_ID, newId=NEW_ID,
                roomsSeparated=True, joinedWorkers=4, payloadSha256=digest(envelope["inputPayload"]),
                envelopeSha256=digest(envelope), productionWrites=0, publicPublishes=0,
                realRefunds=0, externalBusinessMessages=0)
    io.print("PRECHECK_OK rooms_separated=true joined_workers=4")
    if not apply:
        return plan
    io.claim(plan)  # Durable exclusive marker BEFORE any mutation; never auto-repeat.
    if old["status"] != "paused":
        io.pause(OLD_ID)
    ensure(io.project(OLD_ID).get("status") == "paused", "P5_PAUSE_READBACK_FAILED")
    io.print("P5_PAUSED_READBACK_OK")
    io.record("prepared-envelope.json", envelope)
    body = {"msgtype": "m.text", "body": "JUCHANG room-corrected run. Preserve p5 failure; no business writes.\n" + leader + "\nJUCHANG_DSH_PROJECT: " + encode(envelope), "m.mentions": {"user_ids": [leader]}}
    sent = io.send(team["leaderDMRoomID"], "juchang-roomfix-" + NEW_ID + "-once", body)
    ensure(isinstance(sent.get("event_id"), str) and sent["event_id"].startswith("$"), "SEND_RESULT_UNKNOWN_DO_NOT_REPEAT")
    record = dict(plan, sentAt=now(), eventId=sent["event_id"], state="sent_not_completed", instruction="Read exact p6 Project/Taskflow; never infer completion from this record.")
    io.record("sent.json", record)
    io.print("DISPATCH_SENT project=" + NEW_ID + " event_prefix=" + sent["event_id"][:12])
    io.print("NOT_COMPLETED: wait for official four-task and Leader terminal readback.")
    return plan


def cli(args, token, optional=False):
    env = dict(os.environ, AGENTTEAMS_AUTH_TOKEN=token)
    try:
        result = subprocess.run(["agt"] + args, env=env, stdin=subprocess.DEVNULL,
                                stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                                timeout=35, encoding="utf-8", errors="replace", check=False)
    except (OSError, subprocess.TimeoutExpired):
        raise RepairError("AGT_EXECUTION_FAILED") from None
    if result.returncode:
        if optional and re.search(r"HTTP 404\b", result.stderr):
            return None
        raise RepairError("AGT_" + "_".join(args[:2]).upper() + "_FAILED")
    # Official projectWrite can return plain 'ok'. Pause still requires JSON readback.
    if args[:2] == ["project", "pause"] and result.stdout.strip() == "ok":
        return {}
    try:
        return json.loads(result.stdout)
    except (TypeError, ValueError):
        raise RepairError("AGT_RESPONSE_INVALID") from None


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None  # Never forward credentials or repeat a PUT after redirects.


class ControllerIO:
    def __init__(self, state_dir):
        self.state_dir = Path(state_dir)
        self.token = ""
        self.sender_id = ""
        self.opener = urllib.request.build_opener(urllib.request.ProxyHandler({}), NoRedirect())

    def api(self, method, endpoint, body=None, auth=True):
        headers = {"Content-Type": "application/json"}
        if auth:
            headers["Authorization"] = "Bearer " + self.token
        request = urllib.request.Request("http://127.0.0.1:6167" + endpoint,
                                         data=None if body is None else encode(body).encode("utf-8"),
                                         headers=headers, method=method)
        try:
            with self.opener.open(request, timeout=15) as response:
                return json.load(response)
        except urllib.error.HTTPError as error:
            raise RepairError("MATRIX_HTTP_" + str(error.code)) from None
        except (OSError, urllib.error.URLError):
            raise RepairError("MATRIX_REQUEST_UNCERTAIN_DO_NOT_REPEAT") from None
        except ValueError:
            raise RepairError("MATRIX_RESPONSE_INVALID") from None

    def login(self):
        password = os.environ.get("AGENTTEAMS_ADMIN_PASSWORD", "")
        ensure(bool(password), "SERVER_ADMIN_CREDENTIAL_UNAVAILABLE")
        login = self.api("POST", "/_matrix/client/v3/login", {"type": "m.login.password", "identifier": {"type": "m.id.user", "user": "admin"}, "password": password}, auth=False)
        self.token = login.get("access_token", "")
        ensure(isinstance(self.token, str) and bool(self.token), "MATRIX_LOGIN_FAILED")
        who = self.api("GET", "/_matrix/client/v3/account/whoami")
        self.sender_id = who.get("user_id", "")
        ensure(self.sender_id.startswith("@admin:"), "SENDER_NOT_ADMIN")

    def logout(self):
        if self.token:
            try:
                self.api("POST", "/_matrix/client/v3/logout", {})
            except Exception:
                self.print("SESSION_LOGOUT_UNCONFIRMED")
        self.token = ""

    def team(self):
        return cli(["get", "teams", TEAM, "-o", "json"], self.token)

    def leader(self, name):
        return cli(["get", "workers", name, "-o", "json"], self.token).get("matrixUserID")

    def members(self, room):
        data = self.api("GET", "/_matrix/client/v3/rooms/" + urllib.parse.quote(room, safe="") + "/members?membership=join")
        ensure(isinstance(data.get("chunk"), list), "MEMBERSHIP_RESPONSE_INVALID")
        return [e["state_key"] for e in data["chunk"] if e.get("type") == "m.room.member" and e.get("content", {}).get("membership") == "join"]

    def project(self, project_id, optional=False):
        return cli(["get", "projects", project_id, "-o", "json"], self.token, optional)

    def pause(self, project_id):
        return cli(["project", "pause", project_id, "--reason", "Authorized room mismatch repair; preserve p5 evidence"], self.token)

    def claim(self, value):
        self.record("attempt.json", dict(value, at=now()))

    def record(self, name, value):
        # Do not overwrite attempt, failed-run artifacts, or uncertain-send records.
        with open(self.state_dir / name, "x", encoding="utf-8") as file:
            file.write(encode(value))
            file.flush()
            os.fsync(file.fileno())

    def send(self, room, txn, body):
        endpoint = "/_matrix/client/v3/rooms/" + urllib.parse.quote(room, safe="") + "/send/m.room.message/" + urllib.parse.quote(txn, safe="")
        return self.api("PUT", endpoint, body)

    @staticmethod
    def print(line):
        print(line, flush=True)


def main(args):
    ensure(len(args) == 3 and args[0] in ("--plan", "--apply"), "USAGE_PLAN_OR_APPLY_SNAPSHOT_STATE_DIR")
    mode, snapshot_file, state_dir = args
    os.umask(0o077)
    with open(snapshot_file, encoding="utf-8") as file:
        snapshot = json.load(file)
    state = Path(state_dir)
    state.mkdir(parents=True, exist_ok=True, mode=0o700)
    ensure(mode != "--apply" or not (state / "attempt.json").exists(), "ATTEMPT_EXISTS_NO_RESEND")
    io = ControllerIO(state)
    try:
        io.login()
        repair_once(io, snapshot, mode == "--apply")
        if mode == "--apply":
            for _ in range(6):
                time.sleep(3)
                try:
                    project = io.project(NEW_ID, optional=True)
                except RepairError:
                    io.print("READBACK_UNAVAILABLE_SENT_NOT_RETRIED")
                    break
                if project is None:
                    continue
                nodes = project.get("nodes", project.get("tasks", []))
                states = [{"task": str(n.get("id", n.get("task_id", n.get("taskId", ""))))[-2:], "status": n.get("status")} for n in nodes]
                io.print("P6_OFFICIAL_READBACK " + encode({"status": project.get("status"), "tasks": states}))
                if any(n.get("status") in ("assigned", "in_progress", "submitted", "completed") for n in nodes):
                    break
    finally:
        io.logout()


if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except Exception as error:
        code = str(error) if isinstance(error, RepairError) and re.fullmatch(r"[A-Z0-9_]+", str(error)) else "LOCAL_OPERATION_FAILED"
        print("STOP: " + code, file=sys.stderr)
        sys.exit(1)
